### PR TITLE
Adding support for cypress tests for server changes

### LINKED ISSUE #5321: Cypress tests for server side change

### PR DESCRIPTION
**Pull request by yatinappsmith**
## Description

This workflow will enable running cypress tests for client and server changes by building the server code available on the branch itself instead of using the Docker image for `release` or `master` branches. This ensures that we can push breaking changes in our code base in a single branch without having to rely on pushing the server changes first & then the client changes.

Fixes #5321

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I have tested the new workflow in my local fork.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/workflows/server-test.yml [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+          
+      - name: Build package
+        env:
+          APPSMITH_MONGODB_URI: "mongodb://localhost:27017/mobtools"
+          APPSMITH_REDIS_URL: "redis://127.0.0.1:6379"
+          APPSMITH_ENCRYPTION_PASSWORD: "password"
+          APPSMITH_ENCRYPTION_SALT: "salt"
+          APPSMITH_IS_SELF_HOSTED: false
+        working-directory: app/server
+        if: steps.filterserver.outputs.server == 'false'
```
Comment (mohanarpit) [inline .github/workflows/server-test.yml:224]: QQ: What does this condition mean? Does it mean that the `mvn package` will run if there are no changes to the server/client code? Or the opposite?

File: .github/workflows/server-test.yml [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+        run: |
+            mvn package -DskipTests
+            
+      - name: Build and test package
```
Comment (mohanarpit) [inline .github/workflows/server-test.yml:228]: This seems like a duplicate step. We are running `mvn package` twice.

File: .github/workflows/server-test.yml [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+          APPSMITH_REDIS_URL: "redis://127.0.0.1:6379"
+          APPSMITH_ENCRYPTION_PASSWORD: "password"
+          APPSMITH_ENCRYPTION_SALT: "salt"
+          APPSMITH_IS_SELF_HOSTED: false
+        working-directory: app/server
+        if: steps.filterserver.outputs.server == 'true'
+        run: |
+            mvn package
+
+      - name: mkdir plugins
```
Comment (mohanarpit) [inline .github/workflows/server-test.yml:240]: Any reason to not use the `build.sh` script and write the commands separately here?

File: .github/workflows/server-test.yml [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: app/client
+        shell: bash
+    strategy:
+      fail-fast: false
+      matrix:
+        job: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+        #job: [0, 1]
```
Comment (mohanarpit) [inline .github/workflows/server-test.yml:273]: Please remove commented code.

**Comment by mohanarpit** [comment]
Any reason for using a separate build file called `server-test.yml` ? I'm asking because now for pushes to release & master branches, the workflows that will get invoked are:
1. server.yml
2. client-test.yml
3. server-test.yml

We should consolidate and run only of them. This ensures that pushes to release & master branches only execute a single workflow with a single suite of tests.

File: .github/workflows/server-test.yml (hunk 4 above)
Comment (yatinappsmith) [reply .github/workflows/server-test.yml:273]: Removed the commented code

File: .github/workflows/server-test.yml (hunk 3 above)
Comment (yatinappsmith) [reply .github/workflows/server-test.yml:240]: Replaced mvn with build.sh

File: .github/workflows/server-test.yml (hunk 1 above)
Comment (yatinappsmith) [reply .github/workflows/server-test.yml:224]: This condition checks if changes are not under server/ folder

File: .github/workflows/server-test.yml (hunk 2 above)
Comment (yatinappsmith) [reply .github/workflows/server-test.yml:228]: First step is Build package and second step is Build and test package. Only one of the steps would be executed based on the path of files changed.

**Comment by yatinappsmith** [comment]
> Any reason for using a separate build file called `server-test.yml` ? I'm asking because now for pushes to release & master branches, the workflows that will get invoked are:
> 
> 1. server.yml
> 2. client-test.yml
> 3. server-test.yml
> 
> We should consolidate and run only of them. This ensures that pushes to release & master branches only execute a single workflow with a single suite of tests.
Included changes in client-test.yml.

File: .github/workflows/client-test.yml [hunk 5]
```diff
@@ -298,6 +425,13 @@ jobs:
           name: cypress-screenshots-${{ matrix.job }}
           path: app/client/cypress/screenshots/
 
+      # Upload the screenshots as artifacts if there's a failure
+      - uses: actions/upload-artifact@v1
+        if: failure()
+        with:
+          name: cypress-screenshots-${{ matrix.job }}
+          path: app/client/cypress/screenshots/
+
```
Comment (mohanarpit) [inline .github/workflows/client-test.yml:434]: This seems like a duplicate of the previous step. Can be removed.
```suggestion
```

File: .github/workflows/client-test.yml [hunk 6]
```diff
@@ -65,7 +80,7 @@ jobs:
         id: yarn-dep-cache-dir-path
         run: echo "::set-output name=dir::$(yarn cache dir)"
 
-      # Retrieve npm dependencies from cache. After a successful run, these dependencies are cached again
+      # Retrieve npm dependenes from cache. After a successful run, these dependencies are cached again
```
Comment (mohanarpit) [inline .github/workflows/client-test.yml:83]:
```suggestion
      # Retrieve npm dependencies from cache. After a successful run, these dependencies are cached again
```

**Review by mohanarpit — APPROVED** [review]

**Pull request merged**